wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/DevfileEnvironmentFactoryTest.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static io.fabric8.kubernetes.client.utils.Serialization.unmarshal;
+import static org.eclipse.che.api.devfile.server.Constants.EDITOR_TOOL_TYPE;
+import static org.eclipse.che.api.devfile.server.Constants.KUBERNETES_TOOL_TYPE;
+import static org.eclipse.che.api.devfile.server.Constants.OPENSHIFT_TOOL_TYPE;
+import static org.eclipse.che.api.devfile.server.DevfileEnvironmentFactory.DEFAULT_RECIPE_CONTENT_TYPE;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+import org.mockito.InjectMocks;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+import org.testng.reporters.Files;
+
+@Listeners(MockitoTestNGListener.class)
+public class DevfileEnvironmentFactoryTest {
+
+  public static final String LOCAL_FILENAME = "local.yaml";
+  public static final String TOOL_NAME = "foo";
+
+  @InjectMocks private DevfileEnvironmentFactory factory;
+
+  @Test(
+      expectedExceptions = IllegalArgumentException.class,
+      expectedExceptionsMessageRegExp =
+          "Unable to process tool '"
+              + TOOL_NAME
+              + "' of type '"
+              + KUBERNETES_TOOL_TYPE
+              + "' since there is no content provider supplied.")
+  public void shouldThrowExceptionWhenRecipeToolIsPresentAndNoURLComposerGiven() throws Exception {
```

Comment:
What's a URLComposer?

Reply:
yep, outdated -- should be contenp provider now, fixing.